### ISSUE TITLE
Can't add Netscaler VPX

### ISSUE THREAD

**Issue by BenoitLair** [apache/cloudstack#5876]
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
* API
* UI
* NETWORK PROVIDER

##### CLOUDSTACK VERSION
 * 4.16
 * 4.15.x

##### CONFIGURATION
* Advanced zone, physical network with Netscaler service provider in section Network Service Providers

##### OS / ENVIRONMENT
* Citrix VPX 13
* VPX 12
* VPX 11.0

##### SUMMARY
* Error when adding Netscaler VPX to an advanced zone

##### STEPS TO REPRODUCE
* Installing a VPX Netscaler VM
Configuring NSIP, SNIP a vlan on 1/2 and a SNIP on this vlan binding on 1/2
On an advanced zone, Network Service Providers, Netscaler menu
Add a Netscaler device form with the following entries : 

ip : NSIP
user : nsroot
mdp : pass
type : Netscaler VPX Loadbalancer
public interface : 1/2
private interface 1/1
dedicated : true

##### EXPECTED RESULTS
* This should add the VPX device

##### ACTUAL RESULTS
* This send me an error on all these versions (Error on the GUI) : 
Add Netscaler device
(Netscaler) Failed to verify device type specified when matching with actuall device type due to Netscalar device type specified does not match with the actuall device type.

**Comment by kurushi9000** [comment]
In plugins/network-elements/netscaler/src/main/java/com/cloud/network/resource/NetscalerResource.java

`if (_deviceName.equalsIgnoreCase("NetscalerMPXLoadBalancer") && nsHw.get_hwdescription().contains("MPX") ||
                            _deviceName.equalsIgnoreCase("NetscalerVPXLoadBalancer") && nsHw.get_hwdescription().contains("NetScaler Virtual Appliance")) {
                        return;
                    }
                    throw new ExecutionException("Netscalar device type specified does not match with the actuall device type.");
                }

Netscaler 13 per example shows : "Netscaler Remote Licensed Virtual Appliance 450000"
This could explain why adding VPX is
bugged.
Perhaps are there other lines causing a problem

**Comment by rohityadavcloud** [comment]
Thanks @kurushi9000 for reporting, the community doesn't have access to netscaler device and the old netscaler plugin isn't maintained for years - you should contact the vendor (Citrix) to help maintain the network plugin or get them to sponsor test infrastructure for the same.

**Comment by kurushi9000** [comment]
Hello @rohityadavcloud , i tested creating a new account on citrix.com and saw Vpx is free and available for downloading
Perhaps this way, the community could access the appliance ?

**Comment by kurushi9000** [comment]
I have tested on VPX 13 freemium
Now i have tested on VPX 11.0 licensed and it returns : "Failed to log in to Netscaler device at a.b.c.d due to Certificates do not conform to algorithm constraints"

**Comment by kurushi9000** [comment]
@sureshanaparti , this is well working with vpx12 and vpx13
The only issue with VPX is when testing type of Netscaler
I successed to add vpx12 and vpx13 to ACS 4.16 👍 
First add add a MIP VPX11.0 (with a developpement licence : [https://www.citrix.com/downloads/netscaler-adc/virtual-appliances/netscaler-vpx-developer-edition.html](url))
After what i deleted the MIP and recreated it on VPX12 freemium or VPX13 freemium
All other features are still working on it with ACS

The only bug is about the check with "Netscaler Remote Licensed Virtual Appliance 450000"

On vpx 11.0 ns hardware description give : "NetScaler Virtual Appliance 450000"
On later vpx ns hardware description gives : Netscaler Remote Licensed Virtual Appliance 450000"

This bug could be easily been corrected, no ?

**Comment by BenoitLair** [comment]
If a Netscaler is being added with vpx11, once upgrade is done towards vpx13, default password is forced to change

So when vpx13 password is changed, it need to be changed in ACS Database as there are no Gui for this on ACS for an existing Netscaler device
Then new password has to be encrypted with Jasypt library with two way handshake using the private "key" given on ACS cloudstack installation (cloudstack-setup-databases cloud:<password>@<DbIp> --deploy-as=<Db_User:<Db_Pass> -e file -m <key> -k <key> -i <Acs_ip_node>)

There is an online encryption tool here : [https://www.devglan.com/online-tools/jasypt-online-encryption-decryption](url)

The encrypted password to be changed is on host_details, entry named 'password' where "host_details"."host_id" is "external_load_balancer_devices"."host_id"

Created an VPC Redundant Offering with Netscaler public LB capability .
Lb rules are well working !

Tested on vpx 11.0, vpx 12.1 and vpx 13.0 build 84.11

**Comment by DaanHoogland** [comment]
@BenoitLair is there any work on this going on?

**Comment by rohityadavcloud** [comment]
It wouldn't be possible to do this as we don't have access to the said component/hardware/appliance. Need more information.

**Comment by BenoitLair** [comment]
Hello Rohit,
Is there no more anybody from citrix in ACS Dev Team ?
I can see with Citrix asking for a Development account with freemium VPX versions
It is a feature kill if we cant add vpx into ACS Mgmt server

Adding Netscaler is not possible due to "Platform" label value checking with value "NetScaler Virtual Appliance 450000" which is now is now "Netscaler Remote Licensed Virtual Appliance 450000"

The test should implement checking LIKE "Netscaler%%Virtual Appliance 450000"
Also we should have choice of editing password for a Netscaler device

Also there are been some minor changes when working with a VPC with External Load balancer of type Netscaler
When adding some LB rules, the API calls had some little changes

For working with it i installed a Nginx mounting NS ip device declared in ACS and forwarding request to a NS MIP ip

I used lua rewrite file and subfilters in order to adapt NS API changes

`
map $http_upgrade $connection_upgrade {
default upgrade;
'' close;
}

upstream tunnel_ns-vpx13-ssl {
		#use MIP created on Netscaler vpx13, does work with vpx12
        server aa.bb.cc.dd:80;
        keepalive 32;
}

proxy_cache_path /var/cache/nginx-vpx13-ssl levels=1:2 keys_zone=ns-vpx13-ssl_cache:10m max_size=3g inactive=120m use_temp_path=off;

server{
listen 443 ssl;
server_name ee.ff.gg.hh; # use NS ip declared on ACS Mgmt server, has been created with vpx11 device due to constraint 'NetScaler Virtual Appliance'

ssl_certificate /etc/ssl/certs/cert-selfsigned.crt;
ssl_certificate_key /etc/ssl/private/cert-selfsigned.key;
ssl_dhparam /etc/ssl/certs/cert-dhparam.pem;

access_log      /var/log/nginx/ns-vpx13-ssl-access.log;
error_log              /var/log/nginx/ns-vpx13-ssl-error.log;

   location / {

		access_by_lua_file /etc/nginx/vpx13-prod-ee.ff.gg.hh.lua;

		sub_filter_types text/html text/css text/xml application/json;
		sub_filter 'NITRO' 'NITRO2';
		sub_filter 'Login Failure' 'Login Failure22';
		sub_filter 'Netscaler Remote Licensed Virtual Appliance' 'NetScaler Virtual Appliance';

		client_max_body_size 50M;
		proxy_set_header Connection "";
		proxy_set_header Host $http_host;
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header X-Forwarded-Proto $scheme;
		proxy_set_header X-Frame-Options SAMEORIGIN;
		proxy_buffers 256 16k;
		proxy_buffer_size 16k;
		proxy_read_timeout 600s;
		proxy_cache ns-vpx13-ssl_cache;
		proxy_cache_revalidate on;
		proxy_cache_min_uses 2;
		proxy_cache_use_stale timeout;
		proxy_cache_lock on;
		proxy_http_version 1.1;
		proxy_pass http://tunnel_ns-vpx13-ssl;
	}

}


## Content of /etc/nginx/vpx13-prod-ee.ff.gg.hh.lua

function remove_user_key()
  ngx.req.read_body()
  -- log the original body so we can compare to the new one later
  local oldbody = ngx.req.get_body_data()
  --log(oldbody)
  -- grab the POST parameters as a table
  local params = ngx.req.get_post_args()

  -- build up the new JSON string
  local newbody = "{"

   for k,v in pairs(params) do
     -- add all the params we want to keep
     if k ~= "serviceType" then
        -- log('adding"..k.." as "..v.." :')
        newbody = newbody..'"'..k..'":"'..v..'",'
     else
        -- log("adding modified serviceType")
        newbody = newbody..'"'..k..'":"'HTTP'",'
     end
   end
  --remove the last trailing comma before closing this off
  newbody = string.sub(newbody, 0, #newbody-1)
  newbody = newbody.."}"

  ngx.req.set_body_data(newbody)
  -- log(newbody)
end

function format_http_vservers_protocol()
  ngx.req.read_body()
  local body = ngx.req.get_body_data()
  -- ngx.log("not adding user_key")
  ngx.log(ngx.NOTICE, "hello world")
  -- if string.find(body,"Cloud-Service-") then
        body = string.gsub(body, "TCP", "HTTP")
        ngx.req.set_body_data(body)
  -- end
end

if ngx.req.get_method() == "POST" then
  -- remove_user_key()
  -- ngx.req.read_body()
  -- local oldbody = ngx.req.get_body_data()
  -- log(oldbody)

  format_http_vservers_protocol()

  -- ngx.req.read_body()
  -- local body = ngx.req.get_body_data()
  -- if body then
  --    body = string.gsub(body, "TCP", "HTTP")
  -- end
  -- ngx.req.set_body_data(body)
end

`

**Comment by BenoitLair** [comment]
This is a blocker for me in order to upgrade from CS 4.16 to next versions of CS

**Comment by sourabh1515-inf** [comment]
Hi 
we are trying to add NS 14.1 to cloud stack 4.19
and getting the same error.
This send me an error on all these versions (Error on the GUI) :
Add Netscaler device
(Netscaler) Failed to verify device type specified when matching with actuall device type due to Netscalar device type specified does not match with the actuall device type.

**Comment by DaanHoogland** [comment]
@sourabh1515-inf , no it wasn't . it was closed as 'stale', as there was no one working on it and the user didn't respond any more. If you have new input, please re-open or create a [new issue](https://github.com/apache/cloudstack/issues/new/choose)
